crates/re_query_cache2/src/range/results.rs

```diff
@@ -0,0 +1,747 @@
+        }
+    }
+
+    /// Range both the indices and data by zipping them together.
+    ///
+    /// Useful for time-based joins (`range_zip`).
+    #[inline]
+    pub fn range_indexed(&self) -> impl Iterator<Item = (&(TimeInt, RowId), &[T])> {
+        let entry_range = self.entry_range();
+        itertools::izip!(
+            self.range_indices(entry_range.clone()),
+            self.range_data(entry_range)
+        )
+    }
+
+    /// Returns the index range that corresponds to the specified `time_range`.
+    ///
+    /// Use the returned range with one of the range iteration methods:
+    /// - [`Self::range_indices`]
+    /// - [`Self::range_data`]
+    /// - [`Self::range_indexed`]
+    ///
+    /// Make sure that the bucket hasn't been modified in-between!
+    ///
+    /// This is `O(2*log(n))`, so make sure to clone the returned range rather than calling this
+    /// multiple times.
+    #[inline]
+    pub fn entry_range(&self) -> Range<usize> {
+        let Some(indices) = self.indices.as_ref() else {
+            return 0..0;
+        };
+
+        // If there's any static data cached, make sure to look for it explicitly.
+        //
+        // Remember: `TimeRange`s can never contain `TimeInt::STATIC`.
+        let static_override = if matches!(indices.front(), Some((TimeInt::STATIC, _))) {
+            TimeInt::STATIC
+        } else {
+            TimeInt::MAX
+        };
+
+        let start_index = indices.partition_point(|(data_time, _)| {
+            *data_time < TimeInt::min(self.time_range.min(), static_override)
+        });
+        let end_index = indices.partition_point(|(data_time, _)| {
+            *data_time <= TimeInt::min(self.time_range.max(), static_override)
+        });
+
+        start_index..end_index
+    }
+}
+
+impl CachedRangeComponentResults {
+    /// Returns the component data as a dense vector.
+    ///
+    /// Returns an error if the component is missing or cannot be deserialized.
+    ///
+    /// Use [`PromiseResult::flatten`] to merge the results of resolving the promise and of
+    /// deserializing the data into a single one, if you don't need the extra flexibility.
+    #[inline]
+    pub fn to_dense<C: Component>(&self, resolver: &PromiseResolver) -> CachedRangeData<'_, C> {
+        // It's tracing the deserialization of an entire range query at once -- it's fine.
+        re_tracing::profile_function!();
+
+        // --- Step 1: try and upsert pending data (write lock) ---
+
+        thread_local! {
+            /// Keeps track of reentrancy counts for the current thread.
+            ///
+            /// Used to detect and prevent potential deadlocks when using the cached APIs in work-stealing
+            /// environments such as Rayon.
+            static REENTERING: RefCell<u32> = const { RefCell::new(0) };
+        }
+
+        REENTERING.with_borrow_mut(|reentering| *reentering = reentering.saturating_add(1));
+
+        if self.time_range == TimeRange::EMPTY {
```

Comment:
```suggestion
        if self.time_range.is_empty() {
```

otherwise we only capture on empty range, but not e.g. the range `42..13`

Reply:
We dont have a `TimeRange::is_empty` but in any case this path is specifically made for manufactured empty results, which are explicitly built with `TimeRange::EMPTY`. I've added a comment.

Reply:
> We dont have a `TimeRange::is_empty`

I believe you have the power to crate one!